Newsletters: add 284 (2024-01-10) (#1457)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2024-01-10-newsletter.md

---
title: 'Bitcoin Optech Newsletter #284'
permalink: /en/newsletters/2024/01/10/
name: 2024-01-10-newsletter
slug: 2024-01-10-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

> - **Discussion about proposed v3 transaction relay policy:**

Reviewer: I might be biased, but perhaps a more suitable title might be "Discussion about LN anchors and v3 proposal." I think the post was originally created to move out-of-scope discussion (especially the criticisms of CPFP and LN anchors in general) away from the PR.

  Antoine Poinsot [posted][poinsot v3] to Delving Bitcoin to

>   foster`discussion about the proposals for [v3 transaction relay

Reviewer: ```suggestion
  foster discussion about the proposals for [v3 transaction relay
```

  policy][topic v3 transaction relay] and [ephemeral anchors][topic
  ephemeral anchors].  The thread appears to have been motivated by
  Peter Todd [posting][todd v3] a critique of v3 relay policy on his
  blog.  We've arbitrarily divided discussion in to several parts:

  - **Frequent use of exogenous fees may risk mining decentralization:**
    an ideal version of the Bitcoin protocol would reward each miner
    proportionately to their hashrate.  The implicit fees paid in
    transactions preserve that property: a miner with 10% of total
    hashrate has a 10% chance of capturing a next-block fee, whereas a
    miner with 1% of hashrate has a 1% chance.  Fees paid outside of
    transactions and directly to miners, called [out-of-band fees][topic out-of-band fees],
    violate that property: a system that pays miners who together control
    over 55% of hashrate has a 99% chance of getting a transaction
    confirmed within 6 blocks<!-- 1 - (1 - 0.55)**6 -->, likely resulting
    in little effort being made to pay small miners of 1% or less

>     hashrate.  If small miners are paid less than large miners, mining

Reviewer: Perhaps:
```suggestion
    hashrate.  If small miners are paid less than large miners in proportion to the blocks they find, mining
```
or
```suggestion
    hashrate.  If smaller miners cannot profit without access to out-of-band fees, mining
```

Since it's expected that less hashrate => less revenue, but it's problematic when out-of-band fees are significant.

---
title: 'Bitcoin Optech Newsletter #284'
permalink: /en/newsletters/2024/01/10/
name: 2024-01-10-newsletter
slug: 2024-01-10-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Discussion about proposed v3 transaction relay policy:**
  Antoine Poinsot [posted][poinsot v3] to Delving Bitcoin to
  foster`discussion about the proposals for [v3 transaction relay
  policy][topic v3 transaction relay] and [ephemeral anchors][topic
  ephemeral anchors].  The thread appears to have been motivated by
  Peter Todd [posting][todd v3] a critique of v3 relay policy on his

>   blog.  We've arbitrarily divided discussion in to several parts:

Reviewer 2: ```suggestion
  blog.  We've arbitrarily divided the discussion into several parts:
```

      anchor outputs] (LN-Anchors), [DLCs][dlc cpfp], and [client-side
      validation][topic client-side validation] allow at least some of
      their onchain transactions to pay fees _exogenously_, meaning the
      fees paid by the core of the transaction can be augmented with
      fees paid using one or more independent UTXOs.  For example, in
      LN-Anchors the commitment transaction includes one output for each
      party to fee bump using [CPFP][topic cpfp] (the child transaction
      spending an extra UTXO) and the HTLC-Success and HTLC-Failure
      transactions (HTLC-X transactions) are partly signed using
      `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` so they can be aggregated
      into a single transaction with at least one extra input to pay
      fees (the extra input being a separate UTXO).

      Focusing on a thought-experiment version of LN that uses [P2TR][topic
      taproot] and proposed ephemeral anchors, Peter Todd argues that
      its dependency on exogenous fees significantly incentivizes paying
      out-of-band fees.  In particular, the unilateral close of a
      channel with no pending payments ([HTLCs][topic htlc]) would

>       allow a large miner acceping out-of-band fees to include twice as

Reviewer 2: ```suggestion
      allow a large miner accepting out-of-band fees to include twice as
```

      close transactions in a block than could be included by a smaller
      miner who only accepted in-band fees paid for through CPFP fee
      bumping.  The large miner could profitably encourage this by
      offering a moderate discount to users paying out of band.  Peter
      Todd calls that a threat to decentralization.

>       The post does suggest some uses of exogenous fees in protocols is

Reviewer 2: ```suggestion
      The post does suggest that some uses of exogenous fees in protocols is
```

      acceptable, so the concern may be about the frequency of their
      expected use and the relative size difference between using them
      and paying out of band.  In other words, frequently occurring
      zero-pending unilateral closes with a 100% overhead would likely
      be considered more of a risk than potentially rarer unilateral
      closes with 20 pending HTLCs where the overhead is less than 10%.

  - **Implications of exogenous fees on safety, scalability, and costs:**
    Peter Todd's post also noted that existing designs such as
    [LN-Anchors][topic anchor outputs] and future designs that use

>     [ephemeral anchors][topic ephemeral anchors] require each user keep

Reviewer 2: ```suggestion
    [ephemeral anchors][topic ephemeral anchors] require each user to keep
```
(or maybe "require that each user keep")

    experience.  Finally, using [CPFP fee bumping][topic cpfp] or
    attaching additional inputs to a transaction to pay fees exogenously
    requires using more block space and paying more transaction fees
    than paying fees directly from a transaction's input value
    (endogenous fees), making it more expensive in theory even if the
    other problems were not a concern.

  - **Ephemeral anchors introduce a new pinning attack:** as described in
    [last week's newsletter][news283 pinning], Peter Todd described a
    minor pinning attack against uses of ephemeral anchors.  For a
    commitment transaction with no pending payments (HTLCs), an
    unprivileged attacker could create a situation when an honest user
    might have to pay 1.5x to 3.7x more in fees to obtain the feerate
    they intended.  However, if the honest user chose to be [patient][harding
    pinning] instead of spending extra fees, the attacker would end up
    paying some or all of the honest user's fees.  Given that
    zero-pending commitment transactions don't have any
    timelock-dependent urgency, many honest users might choose to be
    patient and get their transaction confirmed at the attacker's

>     expense.  The attack does also work when HTLCs are used, but it

Reviewer 2: ```suggestion
    expense.  The attack also works when HTLCs are used, but it
```

    the LN-Penalty commitment transaction at feerates starting at 10
    sats/vbyte and increasing in each version by 10% until a transaction
    paying 1,000 sats/vbyte is signed.  For a commitment transaction
    with no pending payments (HTLCs), his analysis indicates the signing
    time would be about 5 milliseconds.  However, for each HTLC attached
    to the commitment transaction, the number of signatures would
    increase by 50 and the signing time would increase by 5
    milliseconds.  Bastien Teinturier linked to a [previous
    discussion][bolts #1036] he had started about a similar approach.

      Although the idea may work in some situations, Peter Todd's post
      noted that endogenous fees with presigned incremental fee bumps
      were not a satisfactory replacement for exogenous fees in all cases.
      When the delays required for presigning commitment transactions
      containing multiple HTLCs are multiplied by the several hops on a
      typical payment path, the [delay][harding delays] can easily
      become more than a second and, at least in theory, extend to
      delays of more than a minute.  Peter Todd notes that the delay

>       could be reduced to near constant time if the proposed

Reviewer 2: ```suggestion
      could be reduced to near-constant time if the proposed
```

      transactions).  Even with APO, that would naively imply creating
      <i>n<sup>2</sup></i> signatures, although Peter Todd notes that
      the number of signatures could be reduced by assuming the HTLC-X
      transactions would pay a similar feerate to the commitment
      transaction.

      <!-- Using our tx-calc, 1-in, 22-out for 20 HTLC is 1014 vbytes;
           BOLT3 "expected weights" gives worst-case HTLC-X weight of 705
           = 176.25 vbytes, times 20 is 3525, plus 1014 is 4539. Multiply
           everything by 1,000 s/vb to get total sats -->

      There was an unresolved [debate][teinturier fees] about whether
      endogenous fees would result in an excessive amount of capital
      being reserved for fees.  For example, if Alice signs fee
      variants from 10 s/vb to 1,000 s/vb, she must make decisions based
      on the possibility that her counterparty Bob will put the 1,000
      s/vb variant onchain, even if she wouldn't pay that feerate
      herself.  That means she can't accept payments from Bob where he
      spends the money he would need for the 1,000 s/vb variant.  For
      example, a commitment transaction with 20 HTLCs would make 1

>       million sats temporarily unspendable ($450 USD at time of

Reviewer 2: ```suggestion
      million sats temporarily unspendable ($450 USD at the time of
```

      safety.

  - **Overall conclusions:** discussion was ongoing at the time of
    writing.  Peter Todd concluded that "existing usage of anchor
    outputs should be phased out due to the above-mentioned miner
    decentralization risks; new anchor output support should not be
    added to new protocols or Bitcoin Core."  LN developer Rusty Russell
    [posted][russell inline] about using a more efficient form of
    exogenous fees in new protocols to minimize the concern about
    out-of-band fees.  In the Delving Bitcoin thread, other developers
    working on LN, v3 transactions, and ephemeral anchors defended the
    usefulness of anchors and it seemed likely that they would continue
    working on v3-related protocols.  We will provide updates in future
    newsletters if anything significant changes.

- **LN-Symmetry research implementation** Gregory Sanders
  [posted][sanders lns] to Delving Bitcoin about a [proof-of-concept
  implementation][poc lns] he made of the [LN-Symmetry][topic eltoo]
  protocol (originally called _eltoo_) using a software fork of Core
  Lightning.  LN-Symmetry provides bi-directional payment channels that

>   guarantee being able to publish the latest channel state onchain

Reviewer 2: ```suggestion
  guarantee the ability to publish the latest channel state onchain
```

      they use?"
  a6="An attacker would need to send us version messages with
      manipulated timestamps from multiple peers that they control.
      They would need us to make more than 50% of our outbound
      connections to their nodes, which is hard but much easier
      than completely eclipsing the node."
  a6link="https://bitcoincore.reviews/28956#l-100"

  q7="This PR uses the node's local clock as the upper-bound
      block validation time, rather than network-adjusted time.
      Can we be sure that this reduces esoteric attack surfaces,
      rather than increasing them?"
  a7="Discussion ensued with no clear resolution as to whether it's
      easier for an attacker to affect a node's peer set or its
      internal clock (using malware or NTP fakery, for example),
      but most participants agreed that the PR is an improvement."
  a7link="https://bitcoincore.reviews/28956#l-102"

  q8="Does this PR change consensus behavior? If so, is this a
      soft fork, a hard fork, or neither? Why?"
  a8="It's not a consensus change, it's just a network acceptance
      policy change. Even though block timestamp acceptance is

>       not part of consensus, it's [essential][se timestamp accecptance]

Author: @LarryRuane 

I think "block timestamp acceptance" is general enough a concept to include MTP, and MTP is clearly part of consensus.  Maybe this could be phrased a bit differently to clarify that there can be no consensus rule that includes data from outside the block chain, such as data from each user's own clock.

Reviewer 2: Excellent point, pushed a fixup commit (1700347ea4dcc222c4ea1f4725c43c632c5a8381). Let me know if it could be improved further, thanks.

      layer). The variety of these uses shows that the PR doesn't
      just affect block validity, but there are other implications,
      which we need to validate."
  a9link="https://bitcoincore.reviews/28956#l-197"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

<!-- FIXME: harding to updated Tuesday -->

- BTCPay Server

## Notable code and documentation changes

*Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo], and
[Bitcoin Inquisition][bitcoin inquisition repo].*

- [LND #8308][] raises the `min_final_cltv_expiry_delta` from 9 to 18 as
  recommended by BOLT 02 for terminal payments. This value affects external

>   invoices that not supply the `min_final_cltv_expiry` parameter. The change

Reviewer 2: ```suggestion
  invoices that do not supply the `min_final_cltv_expiry` parameter. The change
```
